functional_tests/README.md

```diff
@@ -0,0 +1,303 @@
+of the Django test case, specifically the
+[`LiveServerTestCase`](https://docs.djangoproject.com/en/stable/topics/testing/tools/#django.test.LiveServerTestCase)
+to perform the initial database setup and run a live server on a random port for the tests.
+
+### Database Snapshot and Restore
+
+We are using [DSLR](https://pypi.org/project/dslr/) for fast database snapshots and restores.
+
+After we have used the Django test runner to set up the test database and initialise it by running migrations, we take a
+DSLR snapshot of this clean, initial state. In the test case fixture, post test, we then restore the clean snapshot,
+ensuring each test gets a clean, migrated database, isolated from other tests.
+
+This setup is done in behave fixtures, kept in [behave_fixtures.py](behave_fixtures.py). These fixtures are the
+registered in the behave hooks in the [environment.py](environment.py)
+
+### Playwright
+
+To give the scenario steps access to a playwright page, we set up a Playwright instance along with a browser and
+browser context in the `before_all` hook, so that it is started once at the beginning of the run. In the
+`before_scenario` hook, we then create a Playwright page object to be used by the scenario, passed to through the behave
+context. This page is closed in the `after_scenario` hook, to ensure each scenario has its own separate page object.
+
+If the `PLAYWRIGHT_TRACE` environment variable is enabled, we also start trace recording at the beginning of the run,
+and start a new trace "chunk" for each scenario, so that traces of individual failed scenarios can be saved to files.
+
+## Why Aren't We Using Existing Django Testing Modules?
+
+At first glance it may appear our custom fixtures and database restore mechanism should be unnecessary, as there are
+multiple choices for modules out there which claim to do what we need. We tried these solutions first and ruled them out
+because of various incompatibilities with our app or testing requirements.
+
+### Pytest-BDD
+
+This is built on Pytest, and we decided to move our unit and integration testing away from Pytest because it has
+compatibility issues with our multi-DB configuration, so this would have suffered the same issue.
+
+Live server testing is accomplished with a fixture, which under the hood uses a Django `LiveServerTestCase`. This
+inherits from
+the [
+`TransactionTestCase`](https://docs.djangoproject.com/en/stable/topics/testing/tools/#django.test.TransactionTestCase),
+which causes us serious compatibility issues, as it uses an isolated test database and flushes all data in between
+tests. We have migrations which seed critical data rows, so a flush operation breaks the app.
+
+We tried various workarounds such as using a fixture file to restore the data, but this runs into Wagtail issues, and
+even if it worked it would be non-ideal as that fixture file would have to be kept up to date and recreated when any new
+seeded data is added.
```

Comment:
There's now a custom `TransactionTestCase` in the repo which should fix this :slightly_smiling_face: Hopefully, that means we can use one of these libraries again :crossed_fingers:

Reply:
This might allow us to use the native behaviour, but we would need to craft a version of `LiveServerTestCase` which inherits from the customised `TransactionTestCase`. 

Although the current solution is also a bit custom, it is actually quite simple in implementation, requiring less than 10 SLOC to work and has proven to be very fast, so I would suggest keeping to this solution initially, and we can separately investigate if replacing it with a customised `LiveServerTestCase` is preferable in terms of understandability and performance down the line?